### PR TITLE
Promote init containers to GA

### PR DESCRIPTION
**Pull request by hodovska**
This is proposed for 1.6
PR moves beta proved concept for init containers to stable. Specification of init containers can be now stated under initContainers field in PodSpec/PodTemplateSpec. Specifying init-containers in annotation is still possible, but will be removed in future version.

```release-note
Init containers have graduated to GA and now appear as a field.  The beta annotation value will still be respected and overrides the field value.
```

### CODE REVIEW

File: pkg/api/v1/conversion.go [hunk 1]
```diff
@@ -380,31 +380,7 @@ func Convert_v1_ReplicationControllerSpec_To_api_ReplicationControllerSpec(in *R
 }
 
 func Convert_api_PodStatusResult_To_v1_PodStatusResult(in *api.PodStatusResult, out *PodStatusResult, s conversion.Scope) error {
-	if err := autoConvert_api_PodStatusResult_To_v1_PodStatusResult(in, out, s); err != nil {
-		return err
-	}
-
-	if old := out.Annotations; old != nil {
```
Comment (smarterclayton) [inline pkg/api/v1/conversion.go]: Code that uses init containers (which is mostly using versioned clients now) still needs to respect the beta annotation.  So this code needs to move into a function that you can use to, if the annotation is set, override the value that is provided with the object.  For backwards compatibility with APIs "beta" annotations take precedence over fields.

Comment (hodovska) [inline pkg/api/v1/conversion.go]: Is it true that this file contains only functions in format
``` 
func Convert_v1_Field_To_api_Field(in *Field, out *api.Field, s conversion.Scope) error 
```
where type of Field is defined in types.go?

Mentioned function would need access to Annotations and InitContainers field, that means pointer to Pod. Is there a file where we put this kind of functions? Is it more acceptable not to modify conversions for init containers promotion?

**Comment by derekwaynecarr** [comment]
@hodovska -- can you split the generated code into a separate commit to allow for easier review?

**Comment by hodovska** [comment]
Does also proposal file have to be updated? (https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/container-init.md) Speaking of statements such as "While pod containers are in alpha state..."

File: pkg/api/v1/types_swagger_doc_generated.go [hunk 2]
```diff
@@ -1274,22 +1274,24 @@ var map_PodSpec = map[string]string{
...
 	"subdomain":                     "If specified, the fully qualified Pod hostname will be \"<hostname>.<subdomain>.<pod namespace>.svc.<cluster domain>\". If not specified, the pod will not have a domainname at all.",
+	"initContainers":                "List of initialization containers belonging to the pod. Init containers are executed in order prior to containers being started. If any init container fails, the pod is considered to have failed and is handled according to its restartPolicy. The name for an init container or normal container must be unique among all containers. Init containers may not have Lifecycle actions, Readiness probes, or Liveness probes. The resourceRequirements of an init container are taken into account during scheduling by finding the highest request/limit for each resource type, and then using the max of of that value or the sum of the normal containers. Limits are applied to init containers in a similar fashion. Init containers cannot currently be added or removed. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/containers",
```
Comment (jwforres) [inline pkg/api/v1/types_swagger_doc_generated.go]: noticed a typo, two of's  "the max of of that"

File: pkg/api/v1/conversion.go (hunk 1 above)
Comment (hodovska) [inline pkg/api/v1/conversion.go]: @jwforres pointing out to this question

**Comment by derekwaynecarr** [comment]
The proposals have moved to kubernetes/community repo, so an update will be needed there.

A few additional asks for this PR:

1. i think we need to error if there is a conflict between the value stored as an annotation versus the non-annotated form.  maybe @smarterclayton has some thoughts there, but i think we should do something to handle conflict resolution in this pr.
1. the existing e2e tests all verify the function works when its set via annotation.  i think we need to have test coverage that verifies the function works even if the annotation is not set.  see: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/init_container.go for the tests in question.

**Comment by hodovska** [comment]
Conversion:
 - v1 => internal api
 possible ways to specify initContainers are: metadata.annotations(PodInitContainerStatusesBetaAnnotationKey, PodInitContainerStatusesAnnotationKey) or field spec.initContainers, but when both are filled error is shown and conversion stops. After conversion metadata.annotations values with keys mentioned above are deleted

 - internal api => v1
 value in spec.initContainers is moved to metadata.annotation fields. It has to be removed from spec.initContainers, so following conversion to internal api will be with no errors.


We might use different approach, where conversion stays as is and possible conflict (annotations vs. spec.initContainers) is resolved in validation. For this I will need more information of how to run v1 validation (before conversion to internal api).

**Comment by smarterclayton** [comment]
We generally do not do conflict resolution, we simply treat the annotation
as a complete override.

File: pkg/api/v1/types.go [hunk 3]
```diff
@@ -2216,6 +2209,12 @@ type PodStatus struct {
 	// More info: https://github.com/kubernetes/kubernetes/blob/master/docs/design/resource-qos.md
 	// +optional
 	QOSClass PodQOSClass `json:"qosClass,omitempty" protobuf:"bytes,9,rep,name=qosClass"`
+	// The list has one entry per init container in the manifest. The most recent successful
+	// init container will have ready = true, the most recently started container will have
+	// startTime set.
+	// Init containers are in alpha state and may change without notice.
```
Comment (derekwaynecarr) [inline pkg/api/v1/types.go]: after this pr, this would no longer be the case.  remove this line about alpha state.

File: test/e2e/common/init_container.go [hunk 4]
```diff
@@ -405,4 +409,363 @@ var _ = framework.KubeDescribe("InitContainer", func() {
 		Expect(len(endPod.Status.InitContainerStatuses)).To(Equal(2))
 		Expect(endPod.Status.ContainerStatuses[0].State.Waiting).ToNot(BeNil())
 	})
+
+	//**********************************************************************************************************************
+
+	It("should invoke init containers on a RestartNever pod", func() {
```
Comment (derekwaynecarr) [inline test/e2e/common/init_container.go]: rather than duplicating all the tests, can we factor this differently so we can run the test twice where we say whether or not to set the annotations as a boolean?

**Comment by derekwaynecarr** [comment]
given @smarterclayton comment, I just have a few minor nits than this is good to me.

**Comment by hodovska** [comment]
@derekwaynecarr , please rewiew changes in commit "e2e tests for spec.initContainers". Diff seems horrible, but basically what I did was just what you suggested, every testing part of the four tests is framed by for cycle. 

```golang
<inicialization, pod definition...>

stable := true
for _ := range 2 {
	if !stable {
		framework.Logf("PodSpec: initContainers in metadata.annotation")
		if err := podutil.SetInitContainersAnnotations(pod); err != nil {
			Expect(err).To(BeNil())
		}
	} else {
		framework.Logf("PodSpec: initContainers in spec.initContainers")
	}

<pod creation and test assertions>

        stable = false
}

```

**Comment by erictune** [comment]
Thanks for doing this.  The code looks correct.  

IIUC, when you round trip a pod (v1 -> internal -> v1) with 1.6 code, then you end up with three copies of the init container spec (alpha annotation, beta annotation, and the field).  So a 1.5 or 1.4 or 1.3 kubelet will be able to understand the Pod and run its init container.

**Comment by derekwaynecarr** [comment]
@hodovska -- please rebase, and hopefully we can get this in before being hit by a rebase bug.

**Comment by erictune** [comment]
Ping on this.  I'd really like to see this go in, as it makes it much prettier to write config that uses init containers.

**Comment by hodovska** [comment]
@derekwaynecarr please accept if you thing it is fine, so it can be merged.

**Comment by derekwaynecarr** [comment]
LGTM

@smarterclayton -- agree?

File: pkg/api/types.go [hunk 5]
```diff
@@ -1911,6 +1909,8 @@ type PodSpec struct {
 	// If not specified, the pod will be dispatched by default scheduler.
 	// +optional
 	SchedulerName string
+	// List of initialization containers belonging to the pod.
+	InitContainers []Container
```
Comment (smarterclayton) [inline pkg/api/types.go]: Why was this moved to the bottom?

File: pkg/api/v1/types.go [hunk 6]
```diff
@@ -2096,21 +2096,6 @@ type PodSpec struct {
...
-	// Init containers may not have Lifecycle actions, Readiness probes, or Liveness probes.
-	// The resourceRequirements of an init container are taken into account during scheduling
-	// by finding the highest request/limit for each resource type, and then using the max of
-	// of that value or the sum of the normal containers. Limits are applied to init containers
-	// in a similar fashion.
-	// Init containers cannot currently be added or removed.
-	// Init containers are in alpha state and may change without notice.
-	// Cannot be updated.
-	// More info: http://kubernetes.io/docs/user-guide/containers
-	InitContainers []Container `json:"-"  patchStrategy:"merge" patchMergeKey:"name"`
```
Comment (smarterclayton) [inline pkg/api/v1/types.go]: Leave this where it is.

File: pkg/api/v1/types.go [hunk 7]
```diff
@@ -2294,12 +2293,6 @@ type PodStatus struct {
 	// +optional
 	StartTime *metav1.Time `json:"startTime,omitempty" protobuf:"bytes,7,opt,name=startTime"`
 
-	// The list has one entry per init container in the manifest. The most recent successful
-	// init container will have ready = true, the most recently started container will have
-	// startTime set.
-	// Init containers are in alpha state and may change without notice.
-	// More info: http://kubernetes.io/docs/user-guide/pod-states#container-statuses
-	InitContainerStatuses []ContainerStatus `json:"-"`
```
Comment (smarterclayton) [inline pkg/api/v1/types.go]: Leave this where it is.

**Comment by smarterclayton** [comment]
A couple of comments and rebase and regenerate, and yes, LGTM

/approve

**Comment by smarterclayton** [comment]
Please add the release note to the PR description:

    ```release-note
    ....
    ```

File: pkg/api/types.go (hunk 5 above)
Comment (hodovska) [inline pkg/api/types.go]: @smarterclayton new ID of json field has to be assigned. I found this approach least impacting, since ID's have to be sorted - I didn't intend to influence other fields.

Comment (smarterclayton) [inline pkg/api/types.go]: We don't sort based on id, and ids aren't added by sort.  Order of fields is supposed to be human / use first, machine last.  Move these back.
